connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/ProcessingContext.java

```diff
@@ -168,11 +102,13 @@ public String toString(boolean includeMessage) {
         builder.append("' with class '");
         builder.append(executingClass() == null ? "null" : executingClass().getName());
         builder.append('\'');
-        if (includeMessage && sourceRecord() != null) {
+        T original = original();
+        if (includeMessage && original instanceof SourceRecord) {
```

Comment:
I code-golfed this down for this diff, but I completely agree with you that this instanceof checking is undesirable.

I was going to go even further and move this logic out of the ProcessingContext, and into LogReporter, and then use your abstract class strategy there. I looked at every or nearly every usage of the ProcessingContext during the course of the refactor, and I never saw the `toString()` function used, only the `toString(boolean)` function, so I think we can change or remove the toString() function without changing publicly visible behavior.

Reply:
Ill expand on this, since it also provides context for the strange casting in DeadLetterQueueReporter.

I think the strange "if source do <x>, if sink do <y>" in `ProcessingContext#toString(boolean)` and `DeadLetterQueueReporter#report()` are both symptoms of the inadequate ErrorReporter signature. Since there's only one ErrorReporter signature used by both sources and sinks, there isn't anything in the type definitions to force the DeadLetterQueueReporter to only work with sinks, so it has to include a runtime check in the implementation.

Similarly, there are two different "kinds" of LogReporter, the one for sources that hits the first branch in the toString, and one that hits the second. Each LogReporter instance only ever takes one of the branches, but that isn't clear from the implementation. If we make ErrorReporter generic, we can have the implementations give functionality for the specific record types when compiling or instantiating, instead of in the report() function.

But pulling on this thread unravelled a bit more: To make ErrorReporter generic, we have to make RetryWithToleranceOperator generic, make LogReporter generic, change some RetryWithToleranceOperator instance reuse, and move the RetryWithToleranceOperator out of WorkerTask. All of which are just lateral refactors, and I think would be easier to address in a follow-up.

Reply:
Thank you for the exhaustive analysis 🙏

I wholeheartedly agree that this would be better suited for a follow-up, and if/when that happens, the comments you've left have created an excellent place to start from.